windows/src/engine/kmcomapi/com/system/keymancontrol.pas

```diff
@@ -760,4 +732,56 @@ procedure TKeymanControl.StartKeyman32;
 end;
 {$ENDIF}
 
+{ TRefreshThread}
+
+constructor TRefreshThread.Create(ALastRefreshToken: Integer);
+begin
+  inherited Create(True);
+  FreeOnTerminate := True;
+  FLastRefreshToken := ALastRefreshToken;
+end;
+
+procedure TRefreshThread.Execute;
+const
+  KR_REQUEST_REFRESH = 0;
+  KR_SETTINGS_CHANGED = 3;
+var
+  msg: TMsg;
+  wm_keyman_refresh: UINT;
+  RefreshHandle: THandle;
+begin
+  // This convoluted way of refreshing keyman ensures that km is init for the thread.
+  // Other methods would work but this is easiest
+
+  wm_keyman_refresh := RegisterWindowMessage('WM_KEYMANREFRESH');
+  RefreshHandle := AllocateHWnd(nil);
+
+  // We currently have two announcements because the KR_REQUEST_REFRESH
+  // announcement is actioned only when Keyman Engine is running, while we
+  // still need to tell any apps listening for changes that the settings
+  // have changed. In future, we should probably refactor this to use a
+  // single broadcast.
+
+  // We use a random number here to so that multiple processes can generate
+  // hopefully unique tokens for refresh, and we share this token with our
+  // current consumer so they can ignore notifications that they have
+  // generated
+
+  // First, post out a KR_SETTINGS_CHANGED for the benefit of Keyman
+  // Configuration and other apps that want it
+  PostMessage(HWND_BROADCAST, wm_keyman_refresh, KR_SETTINGS_CHANGED, FLastRefreshToken);
+
+  // Then, post out a refresh to Keyman Engine and process it in this
+  // thread so that Keyman Engine will grab it and broadcast it, if it is
+  // currently running
+  PostMessage(RefreshHandle, wm_keyman_refresh, KR_REQUEST_REFRESH, 0);
+
+  // Flush the queue but don't stall the thread
+  while PeekMessage(msg, 0, 0, 0, PM_REMOVE) do
+  begin
+    DispatchMessage(msg);
+  end;
```

Comment:
> 3. Most important, the refresh thread now flushes all messages in its queue, as there may be multiple messages relating to the window construction that need to be processed before the wm_keyman_refresh message can be handled. This greatly improves the chance of the notification actually being sent!

So... we flush the message queue _after_ posting the refresh message?  I assume this means that the `PostMessage` line seen above posts to the same queue that `PeekMessage` and `DispatchMessage` draw from?

I'll admit, it has been a while since I last programmed for the Windows message loop, so a little clarification would help.  That said, at a minimum, it's a definite improvement over the previous version, which didn't even consider the need to loop and process any as-of-yet-unprocessed messages that may have remained within the message queue.

Reply:
Yes, this is a little trick. We create our own thread with its own message queue. Under the covers, Keyman Engine attaches to this message queue with a GetMessage hook, so posting to the thread allows us to tell Keyman Engine to do its refresh stuff without needing to load the library directly.

We could refactor this to instead load the library, call a new API endpoint called `Keyman_RequestRefresh()` or something, and then free the library again. That would work also -- and would probably be better in the long run, but I don't want to do that just now because 'beta' 😉